core/templates/dev/head/domain/objects/NumberWithUnitsObjectFactorySpec.js

```diff
@@ -0,0 +1,57 @@
+describe('NumberWithUnitsObjectFactory', function() {
+  beforeEach(module('oppia'));
+
+  describe('number with units object factory', function() {
+    var NumberWithUnits = null;
+    var Units = null;
+    var Fraction = null;
+
+    beforeEach(inject(function($injector) {
+      NumberWithUnits = $injector.get('NumberWithUnitsObjectFactory');
+      Units = $injector.get('UnitsObjectFactory');
+      Fraction = $injector.get('FractionObjectFactory');
+    }));
+
+    it('should convert units to dict format', function() {
+      expect(new Units('kg / kg^2 K mol / (N m s^2) K s').toDict()).toEqual(
+        {kg: -1, K: 2, mol: 1, N: -1, m: -1, s: -1});
+      expect(new Units('mol/(kg / (N m / s^2)').toDict()).toEqual(
+        {mol: 1, kg: -1, N: 1, m: 1, s: -2});
+    });
+
+    it('should convert units from dict to string format', function() {
+      expect(Units.fromDictToString(
+        {kg: -1, K: 2, mol: 1, N: -1, m: -1, s: -1})).toBe(
+          'kg^-1 K^2 mol^1 N^-1 m^-1 s^-1');
+      expect(Units.fromDictToString(
+        {mol: 1, kg: -1, N: 1, m: 1, s: -2, K: -1})).toBe(
+          'mol^1 kg^-1 N^1 m^1 s^-2 K^-1');
+    });
+
+    it('should parse valid units strings', function() {
+      expect(Units.fromRawInputString('kg / (K mol^-2)').toDict()).toEqual(
+        new Units('kg / (K mol^-2)').toDict());
+      expect(Units.fromRawInputString('kg / (K mol^-2) N / m^2').toDict(
+        )).toEqual(new Units('kg / (K mol^-2) N / m^2').toDict());
+    });
+
+    it('should parse valid number with units strings', function() {
+      expect(NumberWithUnits.fromRawInputString('2.02 kg / m^3')).toEqual(
+        new NumberWithUnits('real', 2.02, '', Units.fromRawInputString(
```

Comment:
Don't use other object factories. Use facts. So whatever you write here should be concrete and not output of some other service.

Reply:
Here in the Spec file we've both object factory for NumberWithUnits and Units. The use of fractions and unit factory is required to make their corresponding objects and finally store in 'NumberWithUnits' object.

Reply:
Yes, for creating objects you can use object factory but don't use the method `fromRawInputString`. Create object directly.